File: javascript/20. Valid Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */

var isOpening = function(ch) {
    if( ch == '(' || ch == '[' || ch == '{') {
       return true;
    }
    return false;

}

var isMatching = function( a, b) {
    if( a == '(' && b == ')') {
        return true;
    } else if( a == '[' && b == ']') {
        return true;
    } else if( a == '{' && b == '}') {
        return true;
    } else {
        return false

    }
}
var isValid = function(s) {
    var stack = [];
    
    for( let i = 0 ; i < s.length ; i++ ) {
        if(isOpening(s[i]) ) {
            stack.push(s[i]);
        } else {
            if(stack.length ) {
                if(isMatching(stack[stack.length - 1], s[i] ) ) {
                    stack.pop();
                } else return false;
            } else return false;
        }
    }
    return stack.length == 0;
};

File: Java/20. Valid Parentheses.java
class Solution {
    public boolean isOpening(char ch) {
        return ch == '(' || ch == '{' || ch == '[';
    }
                                                  
    public boolean isMatching( char ch1, char ch2) {
        if( ch1 == '(' && ch2 == ')') return true;
        if( ch1 == '{' && ch2 == '}') return true;
        if( ch1 == '[' && ch2 == ']') return true;
        return false;
    }
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        for(int i = 0 ; i < s.length(); i++ ) {
            char ch = s.charAt(i);
            if( isOpening(ch) ) {
                st.push(ch);
            } else {
                if( st.isEmpty() ) {
                    return false;
                }
                if( !isMatching(st.peek(), ch)) {
                    return false;
                }
                st.pop();
                
            }
           
        }
        return st.isEmpty();
    }
}
